### ISSUE TITLE
关于消融实验的疑问？

### ISSUE THREAD

**Issue by WateverOk** [LeapLabTHU/InLine#6]
您好，我运行了您的代码，并对inline_swin_t进行了消融实验，但是在去掉局部建模以后，top-1并没有达到文章中的80.2%，而是78%，请问为什么会这样呢？
不过我对完整模型也进行了实验，最后的指标是和文章相同，所以想请教一下原因。
![image](https://github.com/user-attachments/assets/fc6cbb14-0a86-40c4-b544-641260924fe4)

**Comment by tian-qing001** [comment]
Hi @WateverOk, thanks for trying our method.
The results in Table.3 (both linear and inline attention) are obtained with APE, which provides necessary positional information for the model. In the final inline attention module, the local bias term already provides positional information and we omit APE for simplicity. For adding APE to inline attention, please take a look at the code below. 
```python
class InLineAttention(nn.Module):
    def __init__(self, dim, window_size, num_heads, qkv_bias=True, qk_scale=None, attn_drop=0., proj_drop=0., shift_size=0,
                 kernel_func='identity', **kwargs):

        ...
        self.ape = nn.Parameter(torch.zeros(1, window_size[0] * window_size[1], dim))
        trunc_normal_(self.ape, std=.02)

    def forward(self, x, mask=None):
        x = x + self.ape

        b, n, c = x.shape
        h = int(n ** 0.5)
        w = int(n ** 0.5)
        num_heads = self.num_heads
        head_dim = c // num_heads
        qkv = self.qkv(x).reshape(b, n, 3, c).permute(2, 0, 1, 3)

        ...

        x = self.proj(x)
        x = self.proj_drop(x)
        return x
```

**Comment by WateverOk** [comment]
感谢您的回复，想要请教一下，如果这时候加上了位置编码的话，会对最终的性能有影响吗？

**Comment by tian-qing001** [comment]
Hi @WateverOk. Adding APE to the vanilla InLine attention based Swin-T can notably improve the performance, as that model has no positional information. In the final InLine-Swin-T setting, adding APE can bring around 0.1 accuracy gain.

**Comment by WateverOk** [comment]
也就是说，在已开源代码的基础上再添加APE的话，还能带来0.1的精度增益？那为什么代码中不添加APE呢？

**Comment by tian-qing001** [comment]
Thanks for your question. The primary goal of our paper is to analyze the gap between Softmax and linear attention. We focus on validating our analyses and designs, instead of pursuing very high performance. APE (and many other positional encodings like CPE, RoPE) can further improve the performance of our models, but this is not the focus of our work. If more competitive results are desired, I recommend adding APE/CPE/RoPE.

**Comment by WateverOk** [comment]
好的，感谢您的回复
